cpp/dolfin/function/Constant.cpp

```diff
@@ -1,39 +1,37 @@
-// Copyright (C) 2019 Chris Richardson, Michal Habera
+// Copyright (C) 2019 Chris Richardson and Michal Habera
 //
 // This file is part of DOLFIN (https://www.fenicsproject.org)
 //
 // SPDX-License-Identifier:    LGPL-3.0-or-later
 
 #include "Constant.h"
+#include <Eigen/Dense>
 #include <vector>
 
-
 using namespace dolfin;
 using namespace dolfin::function;
 
-
 //-----------------------------------------------------------------------------
-Constant::Constant(std::vector<PetscScalar> value, std::vector<int> shape)
-    : value(value), shape(shape){}
+Constant::Constant(PetscScalar value) : value({value})
+{
+  // Do nothing
+}
 //-----------------------------------------------------------------------------
-/// Initialise with a scalar value
-Constant::Constant(PetscScalar value) : value({value}), shape({1}){}
+Constant::Constant(std::vector<int> shape, std::vector<PetscScalar> value)
+    : shape(shape), value(value)
+{
+  // Do nothing
+}
 //-----------------------------------------------------------------------------
-/// Initialise with a 1D or 2D Array
-Constant::Constant(Eigen::Ref<Eigen::Array<PetscScalar, Eigen::Dynamic, Eigen::Dynamic,
+Constant::Constant(
+    const Eigen::Ref<Eigen::Array<PetscScalar, Eigen::Dynamic, Eigen::Dynamic,
                                   Eigen::RowMajor>>
-              array)
-    : shape({(int)array.rows(), (int)array.cols()})
+        c)
+    : shape({(int)c.rows(), (int)c.cols()}), value(c.rows() * c.cols())
 {
-  value.resize(array.rows() * array.cols());
-
-  // Remove trailing 1 in shape for Eigen::Vector (1D array)
-  if (array.cols() == 1)
-    shape.pop_back();
-
   // Copy data from Eigen::Array to flattened vector
-  Eigen::Map<Eigen::Array<PetscScalar, Eigen::Dynamic, Eigen::Dynamic,
-                          Eigen::RowMajor>>
-      v(value.data(), array.rows(), array.cols());
-  v = array;
+  for (int i = 0; i < c.rows(); ++i)
+    for (int j = 0; j < c.cols(); ++j)
+      value[i * c.cols() + j] = c(i, j);
```

Comment:
This may miss some vectorisation optimisations in Eigen.

Reply:
The costs of this is trivial.